source/Calamari.Common/Plumbing/Pipeline/PipelineCommand.cs

```diff
@@ -0,0 +1,129 @@
+﻿#if !NET40
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Autofac;
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Behaviours;
+using Calamari.Common.Features.Deployment;
+using Calamari.Common.Plumbing.Extensions;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Variables;
+
+namespace Calamari.Common.Plumbing.Pipeline
+{
+    public abstract class PipelineCommand
+    {
+        protected virtual IEnumerable<IBeforePackageExtractionBehaviour> BeforePackageExtraction(BeforePackageExtractionResolver resolver)
+        {
+            return Enumerable.Empty<IBeforePackageExtractionBehaviour>();
+        }
+
+        protected virtual IEnumerable<IAfterPackageExtractionBehaviour> AfterPackageExtraction(AfterPackageExtractionResolver resolver)
+        {
+            return Enumerable.Empty<IAfterPackageExtractionBehaviour>();
+        }
+
+        protected virtual IEnumerable<IPreDeployBehaviour> PreDeploy(PreDeployResolver resolver)
+        {
+            return Enumerable.Empty<IPreDeployBehaviour>();
+        }
+
+        protected abstract IEnumerable<IDeployBehaviour> Deploy(DeployResolver resolver);
+
+        protected virtual IEnumerable<IPostDeployBehaviour> PostDeploy(PostDeployResolver resolver)
+        {
+            return Enumerable.Empty<IPostDeployBehaviour>();
+        }
+
+        public async Task Execute(ILifetimeScope lifetimeScope, IVariables variables)
+        {
+            var pathToPrimaryPackage = variables.GetPathToPrimaryPackage(lifetimeScope.Resolve<ICalamariFileSystem>(), false);
+            var deployment = new RunningDeployment(pathToPrimaryPackage, variables);
+
+            try
+            {
+                foreach (var behaviour in GetBehaviours(lifetimeScope, deployment))
+                {
+                    await behaviour;
+
+                    if (deployment.Variables.GetFlag(KnownVariables.Action.SkipRemainingConventions))
+                    {
+                        break;
+                    }
+                }
+            }
+            catch (Exception installException)
+            {
+                Console.Error.WriteLine("Running rollback behaviours...");
+
+                deployment.Error(installException);
+
+                try
+                {
+                    // Rollback behaviours include tasks like DeployFailed.ps1
+                    await ExecuteBehaviour(deployment, lifetimeScope.Resolve<RollbackScriptBehaviour>());
```

Comment:
I'm pretty sure only `DeployPackageCommand` and `DeployJavaArchiveCommand` call `RollbackScriptConvention`.

Also, `JavaLibraryCommand`, `DeployPackageCommand` and `DeployJavaArchiveCommand` call `FeatureRollbackConvention`.

We probably need to add a hookpoint for rollbacks?

Reply:
We can adjust the pipeline once we tackle these commands

Reply:
If that's the case, I think we should remove the rollback altogether and put in a `//TODO` so it's explicitly missing

Reply:
OK, so this is tied in with the ConfiguredScriptBehaviour